test/e2e/e2e_test.go

```diff
@@ -0,0 +1,409 @@
+One or more Kubernetes major-minor versions to run tests for may be given. If omitted, tests will be run for all supported
+Kubernetes versions.
+
+External storage end-to-end tests require a Kubernetes version-specific testdriver YAML file to be defined. An error is
+returned if no corresponding file is found for a given Kubernetes release.
+
+Examples:
+
+# Run tests for all supported versions:
+e2e.test
+
+# Run tests for 1.16 only:
+e2e.test 1.16
+
+# Run tests for 1.16 and 1.14 (but not 1.15):
+e2e.test 1.16 1.14
+
+# Create cluster with a specific suffix:
+e2e.test --name-suffix=$(git rev-parse --short HEAD)
+
+# Retain cluster after completion of the tests:
+e2e.test --retain
+
+# Use cluster referenced by kubeconfig file instead of using a dynamic cluster:
+e2e.test --kubeconfig=$HOME/.kube/config
+
+# Use a custom driver image:
+e2e.test --driver-image=timoreimann/do-csi-plugin:dev
+
+# Use a custom end-to-end test runner image:
+e2e.test --runner-image=timoreimann/k8s-e2e-test-runner:latest
+
+Options:`)
+		flag.PrintDefaults()
+	}
+}
+
+func TestMain(m *testing.M) {
+	flag.BoolVar(&p.long, "long", false, "Run long tests")
+	flag.StringVar(&p.driverImage, "driver-image", "", "The driver container image to use. Triggers a deployment of the \"latest\"-suffixed development manifest into the cluster if given. Otherwise, the built-in driver of the cluster is used.")
+	flag.StringVar(&p.runnerImage, "runner-image", testRunnerImage, "The end-to-end runner image to use.")
+	flag.StringVar(&p.focus, "focus", "", "A custom ginkgo focus to use for external storage tests. Defaults to running all external tests.")
+	flag.StringVar(&p.kubeconfig, "kubeconfig", "", "The kubeconfig file to use. For DOKS clusters where the kubeconfig has been retrieved via doctl, the DIGITALOCEAN_ACCESS_TOKEN environment variable must be set. If not specified, add-hoc DOKS clusters will be created and cleaned up afterwards for each tested Kubernetes version (unless -retain is also specified).")
+	flag.StringVar(&p.nameSuffix, "name-suffix", "", "A suffix to append to the cluster name. If not specified, a random suffix will be chosen. Ignored if -kubeconfig is specified.")
+	flag.BoolVar(&p.retainClusters, "retain", false, "Retain/keep the created cluster(s). Ignored if -kubeconfig is specified.")
+	flag.Parse()
+
+	p.kubeVersions = flag.Args()
+
+	if p.nameSuffix == "" {
+		p.nameSuffix = rand.String(5)
+	}
+
+	var cancel func()
+	ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	c := make(chan os.Signal, 1)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		cancel()
+	}()
+
+	os.Exit(m.Run())
+}
+
+func TestE2E(t *testing.T) {
+	if !p.long {
+		t.Skip("Skipping test because long mode is not set")
+	}
+
+	token := os.Getenv(envVarDigitalOceanAccessToken)
+
+	if len(p.kubeVersions) == 0 {
+		p.kubeVersions = supportedKubernetesVersions
+	}
+
+	for _, kubeVer := range p.kubeVersions {
+		t.Run(kubeVer, func(t *testing.T) {
```

Comment:
We should `t.Parallel()` here so we can test more than one version simultaneously since we would be creating separate clusters for each.

> As a side note, it might be nice to dynamically determine cluster version in the case of `--kubeconfig` or at the least, disallow multiple kubernetes test versions when passing in `kubeconfig` flag

Reply:
I was somewhat hesitant to parallelize here because it makes the log output all intertwined and impossible to read. In my follow up WIP work, the CI does the parallelization which brings cleanly separated logs automatically.

I do see value though in running things in parallel locally as well. How about me adding another CLI flag to optionally stream the output into a set of log files?

Reply:
I like your suggestions to auto-detect the Kubernetes version and disallow multiple versions in case of a given kubeconfig. If you don't mind though, I'd like to keep those ideas in the backlog for now. My vague anticipation is that once those end-to-end tests run in CI, we won't be doing so many local runs anymore except for smoke testing or fixing tests that failed in CI with a shorter feedback loop.